Add the cloneconfig target, and make the standard defconfig target do
"the right thing" for suse kernels: configure the kernel with the
default configuration that matches the running kernel. Much more recent
kernel sources will still get the SUSE default configuration matching
the running kernel even when the running kernel's config.gz would not
work well anymore.

Index: linux-2.6.4/scripts/kconfig/Makefile
===================================================================
--- linux-2.6.4.orig/scripts/kconfig/Makefile
+++ linux-2.6.4/scripts/kconfig/Makefile
@@ -2,7 +2,8 @@
 # Kernel configuration targets
 # These targets are used from top-level makefile
 
-.PHONY: oldconfig xconfig gconfig menuconfig config silentoldconfig
+.PHONY: oldconfig xconfig gconfig menuconfig config silentoldconfig \
+	cloneconfig
 
 xconfig: $(obj)/qconf
 	$< arch/$(ARCH)/Kconfig
@@ -23,6 +24,16 @@ oldconfig: $(obj)/conf
 silentoldconfig: $(obj)/conf
 	$< -s arch/$(ARCH)/Kconfig
 
+check_CONFIG_IKCONFIG_PROC:
+	$(Q)if ! [ -f /proc/config.gz ]; then				\
+	    echo "Running kernel compiled without CONFIG_IKCONFIG_PROC"; \
+	    false;							\
+	fi
+
+_cloneconfig: check_CONFIG_IKCONFIG_PROC
+	$(Q)gzip -cd /proc/config.gz > $(obj)/.config
+cloneconfig: _cloneconfig oldconfig
+
 .PHONY: randconfig allyesconfig allnoconfig allmodconfig defconfig
 
 randconfig: $(obj)/conf
@@ -37,15 +48,20 @@ allnoconfig: $(obj)/conf
 allmodconfig: $(obj)/conf
 	$< -m arch/$(ARCH)/Kconfig
 
-defconfig: $(obj)/conf
-	$< -d arch/$(ARCH)/Kconfig
-
-%_defconfig: $(obj)/conf
-	$(Q)$< -D arch/$(ARCH)/configs/$@ arch/$(ARCH)/Kconfig
+_defconfig: check_CONFIG_IKCONFIG_PROC
+	$(Q)eval "$$(gzip -cd /proc/config.gz)";			\
+	config=$(srctree)/arch/$(ARCH)/defconfig.$$CONFIG_CFGNAME;	\
+	if ! [ -e $$config ]; then					\
+	    echo "Config file $$config not found";			\
+	    false;							\
+	fi;								\
+	cp $$config $(obj)/.config
+defconfig: _defconfig oldconfig
 
 # Help text used by make help
 help:
 	@echo  '  oldconfig	  - Update current config utilising a line-oriented program'
+	@echo  '  cloneconfig     - Update current config to the contents of /proc/config.gz'
 	@echo  '  menuconfig	  - Update current config utilising a menu based program'
 	@echo  '  xconfig	  - Update current config utilising a QT based front-end'
 	@echo  '  gconfig	  - Update current config utilising a GTK based front-end'
